argilla/src/argilla/records/_io/_datasets.py

```diff
@@ -58,11 +60,65 @@ def _record_dicts_from_datasets(dataset: HFDataset) -> List[Dict[str, Union[str,
         Returns:
             Generator[Dict[str, Union[str, float, int, list]], None, None]: A generator of dictionaries to be passed to DatasetRecords.add or DatasetRecords.update.
         """
-        record_dicts = []
+        media_features = HFDatasetsIO._get_media_features(dataset)
+        if media_features:
+            dataset = HFDatasetsIO._cast_images_as_urls(hf_dataset=dataset, columns=media_features)
         try:
             dataset: IterableDataset = dataset.to_iterable_dataset()
         except AttributeError:
             pass
+        record_dicts = []
         for example in dataset:
             record_dicts.append(example)
         return record_dicts
+
+    @staticmethod
+    def _get_media_features(dataset: "HFDataset") -> List[str]:
+        """Check if the Hugging Face dataset contains image features.
+
+        Parameters:
+            hf_dataset (HFDataset): The Hugging Face dataset to check.
+
+        Returns:
+            bool: True if the Hugging Face dataset contains image features, False otherwise.
+        """
+        media_features = []
+        for name, feature in dataset.features.items():
+            if isinstance(feature, Image):
+                media_features.append(name)
+        return media_features
+
+    @staticmethod
+    def _cast_images_as_urls(hf_dataset: "HFDataset", columns: List[str]) -> "HFDataset":
+        """Cast the image features in the Hugging Face dataset as URLs.
+
+        Parameters:
+            hf_dataset (HFDataset): The Hugging Face dataset to cast.
+            repo_id (str): The ID of the Hugging Face Hub repo.
+
+        Returns:
+            HFDataset: The Hugging Face dataset with image features cast as URLs.
+        """
+
+        unique_identifier = uuid4().hex
+
+        def batch_fn(batch, column):
+            data_uris = []
+            for sample in batch:
+                data_uris.append(pil_to_data_uri(sample))
+            return {f"column_{unique_identifier}": data_uris}
+
+        for column in columns:
+            hf_dataset = hf_dataset.map(
+                function=batch_fn,
+                with_indices=False,
+                batched=True,
+                input_columns=[column],
+                remove_columns=[column],
+                fn_kwargs={"column": column},
+            )
+            hf_dataset = hf_dataset.rename_column(
+                original_column_name=f"column_{unique_identifier}", new_column_name=column
+            )
+
+        return hf_dataset
```

Comment:
Since datasets supports modifying column values directly, what about this approach? 
```suggestion
        def batch_fn(batch, columns: List[str]):
            return {
                column: [pil_to_data_uri(sample[column]) for sample in batch]
                for column in columns 
            }

        return hf_dataset.map(
            function=batch_fn,
            batched=True,
            fn_kwargs={"columns": columns},
        )
```

Reply:
@frascuchon I tested this out and an error is raise when we iterate over. I think this is because the dataset's `Features` need to match the samples.
 
I worked on this for a while trying a few approaches like `Dataset.cast_column` and found that this was the best I could get. 

Happy to try out more though.

Reply:
I thought we could apply the image conversion for all the image columns at once. That was my motivation, but the code is fine as is.